KAFKA-6003: Accept appends on replicas unconditionally when local producer state doesn't exist

apurvam opened the pull request:
Without this patch, if the replica's log was somehow truncated before
the leader's, it is possible for the replica fetcher thread to
continuously throw an OutOfOrderSequenceException because the
incoming sequence would be non-zero and there is no local state.

This patch changes the behavior so that the replica state is updated to
the leader's state if there was no local state for the producer at the
time of the append.

apurvam commented:
cc @hachikuji

hachikuji reviewed (commented):
Thanks for the patch. The only comment I have is that it would be nice if we could consolidate the loading path and the replication path so that the checks are consistent.

apurvam commented:
So I went down the path of trying to consolidate the two paths. Essentially, the current `loadingFromLog` flag is replaced by `shouldUnconditionallyAcceptFirstAppend`. The latter is `true` unless the append is from the client.

The problem is that it seems that duplicates appends are also OK during log recovery. In particular, the following test fails when we validate appends during recovery: https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/log/LogSegmentTest.scala#L322

The reason for the failure is that `DuplicateSequenceException` is raised.. Before my changes, we would never validate appends during log recovery, so we would never check for duplicates. 

In my latest patch, I catch and ignore a `DuplicateSequenceException` during recovering and log it. I think this should suffice. Overall, I prefer the consolidated code patch.

ijuma commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -876,11 +876,12 @@ class Log(@volatile var dir: File,
+    // We accept the first append with a non zero sequence if we are building the producer state from the log or during
+    // an append on the replica (where the local state may be truncated at a different time than the leader, or perhaps
+    // when new replicas are being spun up).
+    val appendInfo = producers.getOrElseUpdate(producerId, producerStateManager.prepareUpdate(producerId, acceptFirstAppendUnconditionally = !isFromClient))
```
Nit: long line.

ijuma commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -142,20 +142,16 @@ private[log] class ProducerIdEntry(val producerId: Long, val batchMetadata: muta
- *                       difference in behavior is that we do not validate the sequence number of the first append
- *                       since we may have lost previous sequence numbers when segments were removed due to log
- *                       retention enforcement.
+ * @param acceptFirstAppendUnconditionally
```
We should document this parameter. Maybe some/all of the code comment could be moved here since it's more generally useful (imagine someone trying to decide if they should set this to true or false or reading code that sets this).

ijuma commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -166,8 +162,18 @@ private[log] class ProducerAppendInfo(val producerId: Long,
+            // append to have a non zero sequence. This should also be accepted.
+            if (!acceptFirstAppendUnconditionally)
+              throw new UnknownProducerIdException(s"Found no record of producerId=$producerId on the broker. It is possible " +
+                s"that the last message with the producerId=$producerId has been removed due to hitting the retention limit.")
```
Nit: maybe "removed due to the retention configuration"?

ijuma commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -204,22 +210,13 @@ private[log] class ProducerAppendInfo(val producerId: Long,
     } else {
-      append(batch.producerEpoch, batch.baseSequence, batch.lastSequence, batch.maxTimestamp, batch.lastOffset,
-        batch.isTransactional)
+      append(batch.producerEpoch, batch.baseSequence, batch.lastSequence, batch.maxTimestamp, batch.lastOffset, batch.isTransactional)
```
Nit: long line.

ijuma commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -204,22 +210,13 @@ private[log] class ProducerAppendInfo(val producerId: Long,
-      // skip validation if this is the first entry when loading from the log. Log retention
-      // will generally have removed the beginning entries from each producer id
-      validateAppend(epoch, firstSeq, lastSeq)
+  def append(epoch: Short, firstSeq: Int, lastSeq: Int, lastTimestamp: Long, lastOffset: Long, isTransactional: Boolean): Unit = {
```
Nit: long line. Generally, the code should fit within the GitHub review window. Otherwise, it's easy to miss things.

ijuma commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -145,15 +146,21 @@ class LogSegment(val log: FileRecords,
+        }
+      } catch {
+        case _ : DuplicateSequenceException =>
+          info(s"Not updating producer state with producerID: ${batch.producerId()}, and base sequence ${batch.baseSequence()}, " +
```
Nit: we typically write "producer id" in log statements. Also, `()` is not needed in either of the two variables.

ijuma reviewed (commented):
I had a quick look and left some nit comments. :)

hachikuji commented:
@apurvam Should duplicates be allowed during recovery? I can't think of a situation where that could occur other than an error.

apurvam commented:
Synced with @hachikuji offline, and we _think_ the test case which caused duplicates during recovery was wrong. Will confirm today and update the code as necessary.

Commit:
Revamped the patch so that we _only_ validate appends which originate
from the client. The problem is that when we have compacted topics, it
is possible for batches to be compacted on the follower but not on the
leader. This case would also lead to an OutOfOrderSequencException
during replication. The same applies to when we rebuild state from
compacted topics: we would get gaps in the sequence numbers, causing the
OutOfOrderSequence.

In general, once the append is validated and written to the leader the
first time, revalidating it is undesirable since we can't do anything if
validation fails, and also because it is hard to maintain the correct
assumptions during validation, leading to spurious validation failures.

apurvam commented:
@hachikuji brought up a good point: when we have compacted topics, we will have gaps in sequence numbers. As a result, validation will fail when loading from the log, and potentially also during replication (depending when compaction kicks in).

So we decided that the best way to deal with all this is to validate the batch only once, when the first append comes from the client. Future appends on the replicas or when rebuilding state are accepted unconditionally. This is desirable since we can't do much if revalidation fails, and more importantly, the guarantees during revalidation are strictly weaker (we can check for duplicates at best).

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -142,16 +142,17 @@ private[log] class ProducerIdEntry(val producerId: Long, val batchMetadata: muta
                                       currentEntry: ProducerIdEntry,
                                       validateSequenceNumbers: Boolean,
-                                      loadingFromLog: Boolean) {
+                                      isFromClient: Boolean) {
```
There's a bit of overlap between `validateSequenceNumbers` and `isFromClient`. The only difference is that `isFromClient` also skips the fencing check. Would it make sense to make the check `validateEpoch` or something like that for consistency?

apurvam replied:
I agree that this is definitely a bit tricky. The problem is that the flags are sort of hierarchical: 

When the request is from the client, but for the consumer offsets topic, we _only_ want to validate the epoch, but not the sequence numbers. In this case, `validateSequenceNumbers` will be false.

When the request isn't from the client, we don't want to do any validation. In this case, it doesn't matter what the value of `validateSequenceNumbers` is.

Not sure if there is a clearly superior way to represent these kinds of hierarchical flags in code, other than creating an enum type like `ValidationType`. This could have values `NoValidation`, `EpochOnlyValidation`, `FullValidation`. 

Update: I tried to implement this, but it's a bit painful with scala enums :(

apurvam replied:
I added a `ValidationType` pseudo-enum (using sealed traits and case objects per @ijuma's suggestion). This makes the code cleaner IMO.

hachikuji commented on core/src/test/scala/unit/kafka/log/ProducerStateManagerTest.scala:
```diff
@@ -321,6 +321,28 @@ class ProducerStateManagerTest extends JUnitSuite {
   }
 
   @Test
+  def testAcceptAppendWithoutProducerStateOnReplica(): Unit = {
```
Could we have a similar test case which has a gap in the sequence numbers and verifies that they are accepted when `isFromClient` is false?

apurvam replied:
ack. Added this test case.

Commit:
Add test case which shows that an OutOfOrderSequencException isn't raised for non-client appends

Commit:
Consolidate the 'validateSequenceNumbers' and 'isFromClient' flags into a 'ValidationType' which has different subtypes. Makes the code eaiser to read

Commit:
Incorporate Ismael's improvement

Commit:
Add vertical whitespace

Commit:
Don't make ValidationType a subtype of BaseEnum. This allows us to drop the 'name' field from the ValidationType, making it more compact

hachikuji commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -138,48 +147,45 @@ private[log] class ProducerIdEntry(val producerId: Long, val batchMetadata: muta
+        throw new ProducerFencedException(s"Producer's epoch is no longer valid. There is probably another producer " +
+          s"with a newer epoch. $producerEpoch (request epoch), ${currentEntry.producerEpoch} (server epoch)")
+      }
+      if (validationType == ValidationType.Full) {
```
One minor suggestion: maybe we can add two methods `validateEpoch` and `validateSequence`. Then we can switch on the validation type:
```scala
validation type match {
  case None =>
  case EpochOnly => 
    checkEpoch()
  case Full => 
    checkEpoch()
    checkSequence()
}
```

apurvam replied:
That's a good suggestion. I have incorporated it.

ijuma replied:
Yeah, best to use pattern matching (with exhaustiveness checks) than if/else.

Commit:
Incorporate Jason's refactor suggestion

hachikuji reviewed (approved):
LGTM. Thanks for the patch.

hachikuji commented:
Merged to trunk and cherry-picked into 1.0.

ijuma commented on core/src/main/scala/kafka/log/ProducerStateManager.scala:
```diff
@@ -138,49 +147,58 @@ private[log] class ProducerIdEntry(val producerId: Long, val batchMetadata: muta
+ * @param validationType Indicates the extent of validation to perform on the appends on this instance. Offset commits
+ *                       coming from the producer should have EpochOnlyValidation. Appends which aren't from a client
+ *                       will not be validated at all, and should be set to NoValidation. All other appends should
+ *                       have FullValidation.
```
Seems like we didn't update the enum names in this comment. My suggestion is to refer to `ValidationType` values for the explanation and include some javadoc for each type. This comment includes the information for `EpochOnly`, we should ideally include some for `None` an `Full`. All the important information is available in the PR, so it's just a matter of moving it to a place that people will see when trying to understand why we don't validate in some cases. Maybe that can be done together with your NPE fix @apurvam. Thanks!

apurvam replied:
Updated the documentation here: https://github.com/apache/kafka/pull/4022/commits/814434f4f91f40f6db9b3e6da08bb61d90265911
